src/common/resource-request/src/lib.rs

```diff
@@ -19,52 +20,59 @@ use serde::{Deserialize, Serialize};
 #[derive(Clone, Debug, PartialEq, Serialize, Deserialize, Default)]
 #[cfg_attr(feature = "python", pyclass(module = "daft.daft"))]
 pub struct ResourceRequest {
-    pub num_cpus: Option<f64>,
-    pub num_gpus: Option<f64>,
-    pub memory_bytes: Option<usize>,
+    num_cpus: Option<f64>,
+    num_gpus: Option<f64>,
+    memory_bytes: Option<usize>,
 }
 
 impl ResourceRequest {
-    #[must_use]
     pub fn new_internal(
         num_cpus: Option<f64>,
         num_gpus: Option<f64>,
         memory_bytes: Option<usize>,
-    ) -> Self {
-        Self {
+    ) -> DaftResult<Self> {
+        if let Some(num_gpus) = num_gpus {
+            if num_gpus < 0.0 {
+                return Err(DaftError::ValueError(format!(
+                    "ResourceRequest num_gpus must be nonnegative, got {}",
+                    num_gpus
+                )));
+            }
+
+            if num_gpus > 1.0 && num_gpus.fract() != 0.0 {
+                return Err(DaftError::ValueError(format!(
+                    "ResourceRequest num_gpus greater than 1 must be an integer, got {}",
+                    num_gpus
+                )));
+            }
+        }
+
+        Ok(Self {
             num_cpus,
             num_gpus,
             memory_bytes,
-        }
+        })
     }
 
     #[must_use]
     pub fn default_cpu() -> Self {
-        Self::new_internal(Some(1.0), None, None)
+        Self::new_internal(Some(1.0), None, None).unwrap()
     }
 
-    #[must_use]
-    pub fn or_num_cpus(&self, num_cpus: Option<f64>) -> Self {
-        Self {
-            num_cpus: self.num_cpus.or(num_cpus),
-            ..self.clone()
-        }
+    pub fn or_num_cpus(&self, num_cpus: Option<f64>) -> DaftResult<Self> {
+        Self::new_internal(self.num_cpus.or(num_cpus), self.num_gpus, self.memory_bytes)
     }
 
-    #[must_use]
```

Comment:
`must_use` requires that the return value of the function is not discarded when called. `Result` actually already includes this so when I changed these methods to return that, I removed the `#[must_use]` on them. Clippy yells at me otherwise